File: renewal-fails-expired-card/node/notify-expired-cards.js
/**
 * Email subscribers whose saved card is expired or revoked, before the renewal fails.
 *
 * A subscription renewal fails when the card on the contract has expired or the payment
 * method was revoked. Instead of waiting for the failed billing attempt, this walks the
 * active contracts, finds the ones whose card is expired or whose method is gone, and
 * sends the built-in email with customerPaymentMethodSendUpdateEmail. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/shopify/renewal-fails-expired-card/
 */
import { pathToFileURL } from "node:url";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function cardExpired(month, year, nowYear, nowMonth) {
  if (month == null || year == null) return false;
  return year < nowYear || (year === nowYear && month < nowMonth);
}

export function needsUpdate(contract, nowYear, nowMonth) {
  if (contract.status !== "ACTIVE") return false;
  const pm = contract.customerPaymentMethod;
  if (!pm) return false;
  if (pm.revokedAt) return true;
  const card = pm.instrument || {};
  return cardExpired(card.expiryMonth, card.expiryYear, nowYear, nowMonth);
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const CONTRACTS_QUERY = `
query($cursor: String) {
  subscriptionContracts(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id status
      customerPaymentMethod {
        id revokedAt
        instrument { ... on CustomerCreditCard { expiryMonth expiryYear } }
      }
    }
  }
}`;

const SEND_EMAIL = `
mutation($id: ID!) {
  customerPaymentMethodSendUpdateEmail(customerPaymentMethodId: $id) {
    customer { id }
    userErrors { field message }
  }
}`;

async function* contracts() {
  let cursor = null;
  while (true) {
    const data = (await gql(CONTRACTS_QUERY, { cursor })).subscriptionContracts;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function sendUpdateEmail(paymentMethodId) {
  const result = (await gql(SEND_EMAIL, { id: paymentMethodId })).customerPaymentMethodSendUpdateEmail;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
}

export async function run() {
  const now = new Date();
  const nowYear = now.getUTCFullYear();
  const nowMonth = now.getUTCMonth() + 1;
  let notified = 0;
  const seen = new Set();
  for await (const contract of contracts()) {
    if (!needsUpdate(contract, nowYear, nowMonth)) continue;
    const pmId = contract.customerPaymentMethod.id;
    if (seen.has(pmId)) continue;
    seen.add(pmId);
    console.warn(`Contract ${contract.id} has an expired or revoked card. ${DRY_RUN ? "would email" : "emailing"}`);
    if (!DRY_RUN) await sendUpdateEmail(pmId);
    notified++;
  }
  console.log(`Done. ${notified} customer(s) ${DRY_RUN ? "to email" : "emailed"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: renewal-fails-expired-card/python/notify_expired_cards.py
"""Email subscribers whose saved card is expired or revoked, before the renewal fails.

A subscription renewal fails when the card on the contract has expired or the payment
method was revoked. Instead of waiting for the failed billing attempt, this walks the
active contracts, finds the ones whose card is expired or whose method is gone, and
sends the built-in update-payment-method email with customerPaymentMethodSendUpdateEmail.
Run on a schedule. Safe to run again and again.
"""
import os
import datetime
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("notify_expired_cards")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CONTRACTS_QUERY = """
query($cursor: String) {
  subscriptionContracts(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    nodes {
      id status
      customerPaymentMethod {
        id revokedAt
        instrument { ... on CustomerCreditCard { expiryMonth expiryYear } }
      }
    }
  }
}"""

SEND_EMAIL = """
mutation($id: ID!) {
  customerPaymentMethodSendUpdateEmail(customerPaymentMethodId: $id) {
    customer { id }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def card_expired(month, year, now_year, now_month):
    """A card is expired once the month after its expiry has started."""
    if month is None or year is None:
        return False
    return (year, month) < (now_year, now_month)


def needs_update(contract, now_year, now_month):
    if contract.get("status") != "ACTIVE":
        return False
    pm = contract.get("customerPaymentMethod")
    if pm is None:
        return False
    if pm.get("revokedAt"):
        return True
    card = pm.get("instrument") or {}
    return card_expired(card.get("expiryMonth"), card.get("expiryYear"), now_year, now_month)


def send_update_email(payment_method_id):
    result = gql(SEND_EMAIL, {"id": payment_method_id})["customerPaymentMethodSendUpdateEmail"]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])


def contracts():
    cursor = None
    while True:
        data = gql(CONTRACTS_QUERY, {"cursor": cursor})["subscriptionContracts"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def run():
    today = datetime.date.today()
    notified = 0
    seen = set()
    for contract in contracts():
        if not needs_update(contract, today.year, today.month):
            continue
        pm_id = contract["customerPaymentMethod"]["id"]
        if pm_id in seen:
            continue
        seen.add(pm_id)
        log.warning("Contract %s has an expired or revoked card. %s",
                    contract["id"], "would email" if DRY_RUN else "emailing")
        if not DRY_RUN:
            send_update_email(pm_id)
        notified += 1
    log.info("Done. %d customer(s) %s.", notified, "to email" if DRY_RUN else "emailed")


if __name__ == "__main__":
    run()
